**Цель работы:**

Осуществить предварительную обработку данных csv-файла, выявить и устранить проблемы в этих данных.

# Загрузка набора данных

Был скачан файл .ipynb, содержащий подробное описание хода лабораторной работы. Он был переименован согласно требованиям. Далее в папку с блокнотом на компьютер был скачан датасет visits.csv. 

### Описание предметной области

Вариант №6

Набор данных: visits.csv

Атрибуты (Данные пользовательских сессий магазина):
1. уникальный идентификатор пользователя 
2. страна пользователя 
3. устройство пользователя 
4. идентификатор рекламного источника, из которого пришел пользователь 
5. дата и время начала сессии 
6. дата и время окончания сессии


### 1.Чтение файла (набора данных)

Для чтения файла была применена команда pd.read_csv с аргументом sep = ';', так как данные в датасете разделены с помощью знака “;”.

In [2]:
# импорт библиотек, чтение файла с помощью pandas
import pandas as pd
df = pd.read_csv('visits.csv', sep = ';')
df

User_Id         Region   Device   Channel     Session_Start  \
0    9,81449E+11  United States   iPhone   organic   01.05.2019 2:36   
1    2,78966E+11  United States   iPhone   organic   01.05.2019 4:46   
2    5,90706E+11  United States      Mac   organic  01.05.2019 14:09   
3    3,26434E+11  United States  Android    TipTop   01.05.2019 0:29   
4    3,49774E+11  United States      Mac   organic   01.05.2019 3:33   
..           ...            ...      ...       ...               ...   
952  5,55066E+11  United States   iPhone    TipTop  04.05.2019 23:32   
953  7,71716E+11  United States  Android    TipTop  04.05.2019 19:32   
954  74352279110  United States   iPhone    TipTop   04.05.2019 0:13   
955  6,08683E+11  United States  Android  FaceBoom   04.05.2019 3:33   
956  8,75326E+11  United States      Mac   organic   04.05.2019 3:01   

          SESSION_End  
0     01.05.2019 2:45  
1     01.05.2019 4:47  
2    01.05.2019 15:32  
3     01.05.2019 0:54  
4     01.05.2019 3:57  
..                ...  
952  04.05.2019 23:38  
953  04.05.2019 19:48  
954   04.05.2019 0:26  
955   04.05.2019 3:46  
956   04.05.2019 3:15  

[957 rows x 6 columns]

### 2. Обзор данных

2.1 Вывод первых 20 строк с помощью метода head.

Для вывода первых 20 строк датафрейма был использован метод head.

In [3]:
df.head(20)

User_Id         Region   Device       Channel     Session_Start  \
0   9,81449E+11  United States   iPhone       organic   01.05.2019 2:36   
1   2,78966E+11  United States   iPhone       organic   01.05.2019 4:46   
2   5,90706E+11  United States      Mac       organic  01.05.2019 14:09   
3   3,26434E+11  United States  Android        TipTop   01.05.2019 0:29   
4   3,49774E+11  United States      Mac       organic   01.05.2019 3:33   
5   43958116050  United States  Android       organic   01.05.2019 9:03   
6   1,85365E+11  United States   iPhone       organic   01.05.2019 9:37   
7   1,00971E+11  United States      Mac        TipTop   01.05.2019 4:39   
8   3,70456E+11  United States   iPhone       organic  01.05.2019 14:44   
9   1,41838E+11  United States      Mac      FaceBoom   01.05.2019 6:20   
10  9,24309E+11  United States   iPhone       organic  01.05.2019 20:03   
11  7,74938E+11  United States   iPhone  MediaTornado  01.05.2019 14:01   
12  2,44879E+11  United States   iPhone  MediaTornado   01.05.2019 1:59   
13  1,57877E+11  United States      Mac       organic  01.05.2019 10:25   
14  5,24629E+11  United States      Mac       organic  01.05.2019 17:06   
15  1,36227E+11  United States       PC       organic  01.05.2019 19:34   
16  6,84979E+11  United States   iPhone        TipTop   01.05.2019 1:51   
17  9,05776E+11  United States   iPhone       organic  01.05.2019 14:26   
18  15174034813  United States   iPhone      FaceBoom  01.05.2019 15:34   
19  1,88246E+11  United States  Android      FaceBoom  01.05.2019 23:09   

         SESSION_End  
0    01.05.2019 2:45  
1    01.05.2019 4:47  
2   01.05.2019 15:32  
3    01.05.2019 0:54  
4    01.05.2019 3:57  
5   01.05.2019 10:08  
6   01.05.2019 10:00  
7    01.05.2019 4:57  
8   01.05.2019 15:41  
9    01.05.2019 6:54  
10  01.05.2019 20:19  
11  01.05.2019 14:29  
12   01.05.2019 2:21  
13  01.05.2019 10:35  
14  01.05.2019 18:05  
15  01.05.2019 19:39  
16   01.05.2019 2:12  
17  01.05.2019 14:35  
18  01.05.2019 15:56  
19  01.05.2019 23:21

2.2 Оценка данных с помощью метода info.

Для оценки данных был использован метод info. Данный метод предоставляет общую информацию о таблице (названия столбцов, типы данных, хранящихся в них, количество ненулевых значений в каждом столбце). 

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 957 entries, 0 to 956
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   User_Id        957 non-null    object
 1   Region         956 non-null    object
 2   Device         955 non-null    object
 3   Channel        957 non-null    object
 4   Session_Start  955 non-null    object
 5   SESSION_End    955 non-null    object
dtypes: object(6)
memory usage: 45.0+ KB


Метод показал, что в датафрейме содержится 6 столбцов строкового типа и 957 строк. Во всех столбцах, кроме User_Id и Channel имеются пропуски.

2.3 Оценка данных с помощью метода describe.

Для следующей оценки был применен метод describe. Метод используется для получения сводной статистики по данным. Для числовых данных он возвращает количество записей, среднее значение, стандартное отклонение, минимальное и максимальное значения и т.д. Для нечисловых данных - количество записей, количество уникальных значений, наиболее часто встречающееся значение (top) и его частоту (freq).  

In [5]:
df.describe()

User_Id         Region  Device  Channel     Session_Start  \
count           957            956     955      957               955   
unique          865              2       6        4               820   
top     1,79085E+11  United States  iPhone  organic  02.05.2019 20:16   
freq              4            955     421      612                 4   

             SESSION_End  
count                955  
unique               836  
top     02.05.2019 22:48  
freq                   3

Анализируя описанные метрики, можно сказать, что в датасете содержится информация о 865 уникальных пользователях из США, большая часть которых использует iPhone и приходит из рекламного источника organic.


 2.4 Оценка названий столбцов

Следующим шагом была проведена оценка названий столбцов. 

In [6]:
df.columns

Index(['User_Id', 'Region', 'Device', 'Channel', 'Session_Start',
       'SESSION_End'],
      dtype='object')

Названия выглядят разношерстно, требуется корректировка – приведение всех названий к нижнему регистру.

In [7]:
# приведение всех названий к нижнему регистру
df.columns = df.columns.str.lower()
df

user_id         region   device   channel     session_start  \
0    9,81449E+11  United States   iPhone   organic   01.05.2019 2:36   
1    2,78966E+11  United States   iPhone   organic   01.05.2019 4:46   
2    5,90706E+11  United States      Mac   organic  01.05.2019 14:09   
3    3,26434E+11  United States  Android    TipTop   01.05.2019 0:29   
4    3,49774E+11  United States      Mac   organic   01.05.2019 3:33   
..           ...            ...      ...       ...               ...   
952  5,55066E+11  United States   iPhone    TipTop  04.05.2019 23:32   
953  7,71716E+11  United States  Android    TipTop  04.05.2019 19:32   
954  74352279110  United States   iPhone    TipTop   04.05.2019 0:13   
955  6,08683E+11  United States  Android  FaceBoom   04.05.2019 3:33   
956  8,75326E+11  United States      Mac   organic   04.05.2019 3:01   

          session_end  
0     01.05.2019 2:45  
1     01.05.2019 4:47  
2    01.05.2019 15:32  
3     01.05.2019 0:54  
4     01.05.2019 3:57  
..                ...  
952  04.05.2019 23:38  
953  04.05.2019 19:48  
954   04.05.2019 0:26  
955   04.05.2019 3:46  
956   04.05.2019 3:15  

[957 rows x 6 columns]

### 3. Проверка пропусков

С помощью команд isna и sum был проведен подсчет количества пропусков.

In [8]:
df.isna().sum()

user_id          0
region           1
device           2
channel          0
session_start    2
session_end      2
dtype: int64

В ячейках содержатся пропуски, значение которых проблематично заменять средним. Кроме того, доля пропусков относительно общего объёма данных достаточно мала. Поэтому было принято решение удалить строки с отсутствующими значениями. Удаление было совершено с помощью команды dropna. Обновление индексации будет сделано во время работы с дубликатами.

In [9]:
df = df.dropna(subset=['region', 'device', 'session_start'])
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 954 entries, 0 to 956
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   user_id        954 non-null    object
 1   region         954 non-null    object
 2   device         954 non-null    object
 3   channel        954 non-null    object
 4   session_start  954 non-null    object
 5   session_end    954 non-null    object
dtypes: object(6)
memory usage: 52.2+ KB


По итогу работы с пропусками осталось 954 строки из 957, полностью заполненных данными.

### 4. Проверка дубликатов

#### Проверка явных дубликатов

В первую очередь была проведена проверка на явные дубликаты – те строки, которые полностью повторяют уже встречавшиеся в датасете.

In [10]:
df[df.duplicated()]

user_id         region device  channel    session_start  \
480  1,79085E+11  United States    Mac  organic  01.05.2019 7:27   
481  1,79085E+11  United States    Mac  organic  01.05.2019 7:27   

         session_end  
480  01.05.2019 7:49  
481  01.05.2019 7:49

Было обнаружено 2 явных дубликата. Их необходимо удалить и обновить индексацию после проделанных манипуляций.

In [11]:
df = df.drop_duplicates().reset_index(drop=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 952 entries, 0 to 951
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   user_id        952 non-null    object
 1   region         952 non-null    object
 2   device         952 non-null    object
 3   channel        952 non-null    object
 4   session_start  952 non-null    object
 5   session_end    952 non-null    object
dtypes: object(6)
memory usage: 44.8+ KB


#### Проверка неявных дубликатов

Следующий шаг – проверка на неявные дубликаты - альтернативные написания одного и того же значения (или ошибочное написание слов). Для этого с помощью метода value_counts выводятся все уникальные значения, содержащиеся в столбце.

In [12]:
df['channel'].value_counts()

channel
organic         608
FaceBoom        149
TipTop          142
MediaTornado     53
Name: count, dtype: int64

В столбце channel не выявлено дубликатов.

In [13]:
df['device'].value_counts()

device
iPhone     420
Mac        241
Android    185
PC         103
IPHONE       2
MAC          1
Name: count, dtype: int64

В столбце device найдены дубликаты IPHONE и MAC, которые необходимо заменить

In [14]:
df['device'] = df['device'].str.lower()
df['device'].value_counts()

device
iphone     422
mac        242
android    185
pc         103
Name: count, dtype: int64

Все значения столбца были приведены к нижнему регистру, дубликаты исчезли.

In [15]:
df['region'].value_counts()

region
United States    951
USA                1
Name: count, dtype: int64

В столбце region было найдено значение USA, по смыслу равное Unined States. Значение USA было заменено с помощью команды replace.

In [16]:
df['region'] = df['region'].replace('USA', 'United States')
df['region'].value_counts()

region
United States    952
Name: count, dtype: int64

Столбцы user_id, session_start, session_end не нуждаются в проверке на неявные дубликаты, так как первый является непрерывным, а следующие – могут содержать повторы в силу анализа в рамках короткого временного промежутка (1 – 4 мая 2019 года).




По итогам осталось 952 строки без явных и неявных дубликатов.


### 5. Провека типов данных

Атрибут dtypes выводит типы значений в столбцах.

In [17]:
df.dtypes

user_id          object
region           object
device           object
channel          object
session_start    object
session_end      object
dtype: object

Все столбцы имеют строковый тип данных, что не совсем корректно. Поле iser_id должно быть целочисленным. Поля session_start и session_end - datetime.

Помимо этого, id записан в экпоненциальной форме, необходимо скорректировать значения идентификатора.

In [18]:
df['user_id'] = df['user_id'].str.replace(",", ".")
df['user_id'] = df['user_id'].astype(float).astype(int)
df

user_id         region   device   channel     session_start  \
0    981449000000  United States   iphone   organic   01.05.2019 2:36   
1    278966000000  United States   iphone   organic   01.05.2019 4:46   
2    590706000000  United States      mac   organic  01.05.2019 14:09   
3    326434000000  United States  android    TipTop   01.05.2019 0:29   
4    349774000000  United States      mac   organic   01.05.2019 3:33   
..            ...            ...      ...       ...               ...   
947  555066000000  United States   iphone    TipTop  04.05.2019 23:32   
948  771716000000  United States  android    TipTop  04.05.2019 19:32   
949   74352279110  United States   iphone    TipTop   04.05.2019 0:13   
950  608683000000  United States  android  FaceBoom   04.05.2019 3:33   
951  875326000000  United States      mac   organic   04.05.2019 3:01   

          session_end  
0     01.05.2019 2:45  
1     01.05.2019 4:47  
2    01.05.2019 15:32  
3     01.05.2019 0:54  
4     01.05.2019 3:57  
..                ...  
947  04.05.2019 23:38  
948  04.05.2019 19:48  
949   04.05.2019 0:26  
950   04.05.2019 3:46  
951   04.05.2019 3:15  

[952 rows x 6 columns]

Для изменения типа данных из строкового в datetime используется функция pd.to_datetime.

In [19]:
df['session_start'] = pd.to_datetime(df['session_start'], format='%d.%m.%Y %H:%M')
df['session_end'] = pd.to_datetime(df['session_end'], format='%d.%m.%Y %H:%M')
df

user_id         region   device   channel       session_start  \
0    981449000000  United States   iphone   organic 2019-05-01 02:36:00   
1    278966000000  United States   iphone   organic 2019-05-01 04:46:00   
2    590706000000  United States      mac   organic 2019-05-01 14:09:00   
3    326434000000  United States  android    TipTop 2019-05-01 00:29:00   
4    349774000000  United States      mac   organic 2019-05-01 03:33:00   
..            ...            ...      ...       ...                 ...   
947  555066000000  United States   iphone    TipTop 2019-05-04 23:32:00   
948  771716000000  United States  android    TipTop 2019-05-04 19:32:00   
949   74352279110  United States   iphone    TipTop 2019-05-04 00:13:00   
950  608683000000  United States  android  FaceBoom 2019-05-04 03:33:00   
951  875326000000  United States      mac   organic 2019-05-04 03:01:00   

            session_end  
0   2019-05-01 02:45:00  
1   2019-05-01 04:47:00  
2   2019-05-01 15:32:00  
3   2019-05-01 00:54:00  
4   2019-05-01 03:57:00  
..                  ...  
947 2019-05-04 23:38:00  
948 2019-05-04 19:48:00  
949 2019-05-04 00:26:00  
950 2019-05-04 03:46:00  
951 2019-05-04 03:15:00  

[952 rows x 6 columns]

Далее была проведена проверка корректности изменения типов данных.

In [20]:
df.dtypes

user_id                   int64
region                   object
device                   object
channel                  object
session_start    datetime64[ns]
session_end      datetime64[ns]
dtype: object


По итогу получился датафрейм с данными корректного типа.

### 6. Группировка данных

#### Задание 1

*`Группировка - region и количество устройств (device )`*

Группировка выполнена с помощью методов groupby и size (возвращает количество строк в группе). 

In [21]:
df.groupby(['region', 'device']).size()

region         device 
United States  android    185
               iphone     422
               mac        242
               pc         103
dtype: int64

Из результатов данной группировки  можно увидеть, что все подключения осуществлялись из США. Лидер по подключениям среди устройств - iPhone, меньше всего пользователей подключалось, используя ПК.

#### Задание 2

*`Группировка - device и количество рекламных источников каждого типа (channel). Создать датафрейм. Переименовать столбец с количеством в “сount”. Отсортировать по убыванию столбца “count” `*

Был создан новый датафрейм new_df. Помимо вышеописанных, используется метод reset_index, который преобразует Series (получен при группировке) в датафрейм. Метод sort_values с параметром ascending=False сортирует данные в порядке убывания.

In [22]:
new_df = df.groupby(['device', 'channel']).size().reset_index(name='count')
new_df.sort_values('count', ascending=False)

device       channel  count
7    iphone       organic    249
11      mac       organic    172
3   android       organic    115
6    iphone        TipTop     78
15       pc       organic     72
4    iphone      FaceBoom     65
8       mac      FaceBoom     36
0   android      FaceBoom     33
5    iphone  MediaTornado     30
2   android        TipTop     26
10      mac        TipTop     24
12       pc      FaceBoom     15
14       pc        TipTop     14
1   android  MediaTornado     11
9       mac  MediaTornado     10
13       pc  MediaTornado      2

Из результатов данной группировки  можно увидеть, что преобладающее количество пользователей подключались через канал organic. Таким образом, данный канал является основным источником притока людей.

#### Задание 3

*`Сводная таблица (pivot_table) - количество пользователей для каждого канала (channel). Отсортировать по убыванию количества`*

Сводная таблица была сделана с помощью команды pivot_table.

In [23]:
df.pivot_table(index='channel', values='user_id', aggfunc='count').sort_values('user_id', ascending=False)

user_id
channel              
organic           608
FaceBoom          149
TipTop            142
MediaTornado       53

По данным, полученным в сводной таблице, можно сказать, что канал organic по числу пришедших пользователей опережает своих конкурентов в несколько раз. Это показывает, как и говорилось ранее, что данный канал является основным источником людей.

#### Задание 4

*`Сводная таблица (pivot_table) - количество пользователей для каждого устройства (device) - строки и канала - столбцы. Отсортировать по возрастанию столбца device.`*

In [24]:
df.pivot_table(index='device', columns='channel', values='user_id', aggfunc='count').sort_values('device', ascending=True)

channel  FaceBoom  MediaTornado  TipTop  organic
device                                          
android        33            11      26      115
iphone         65            30      78      249
mac            36            10      24      172
pc             15             2      14       72

По данным, полученным в сводной таблице, можно сказать, что пользователи США куда чаще посещают данный сайт, используя продукцию Apple (iPhone, Mac) и канал organic. 

### Вывод

В ходе выполнения лабораторной работы был проанализирован набор данных, содержащий информацию о пользовательских сессиях магазина. Были рассмотрены следующие данные: id пользователей, регион, устройство, источник рекламы, а также время сессии.
Были применены следующие методы предобработки:
1)	Получение общей информации и сводной статистики по данным для первичного анализа с помощью методов info и describe ;
2)	Приведение к одному виду названий столбцов, используя нижний регистр за счет метода str.lower();
3)	Подсчет и удаление строк с пропущенными значениями;
4)	drop_duplicates().reset_index(drop=True) – удаление явных дубликатов с обновлением индексации;
5)	Выявление и удаление неявных дубликатов;
6)	Просмотр и изменение типов данных с помощью методов dtype(), astype(), to_datetime().

При выполнении группировок и составлении сводных таблиц были сделаны выводы о том, что пользователи США куда чаще посещают данный сайт, используя продукцию Apple (iPhone, Mac) и канал organic. Соответственно, это является основным источником людей для данного интернет-магазина, что может быть использовано для более удачной рекламной стратегии.



### Дополнительное задание

### Задание №6

**`Добавить столбец - “Длительность сессии” (расчетный). Создать столбец “Категория длительности” (с помощью категоризации). Выделить минимум 3 категории (низкая, высокая, средняя), фильтрацию для уровня длительности выбрать самостоятельно, аргументировать выбор. Создать сводную таблицу: средняя и медианная длительность сессии по категории длительности и каналу. `**

Добавляется столбец "длительность сессии", рассчитывается (в минутах). bins - границы для определения категорий, выбраны из собственных соображений о длительности пребывания на сайте интернет-магазина, labels - метки фильтрации, include_lowest=True - параметр, делающий так, чтобы левая граница метки учитывалась. Создается сводная таблица по условиям, описанным в задании.

In [68]:
df['duration_minute'] = (df['session_end'] - df['session_start']).dt.total_seconds() / 60

bins = [0, 20, 60, float('inf')]
labels = ['низкая', 'средняя', 'высокая']
df['duration_category'] = pd.cut(df['duration_minute'], bins=bins, labels=labels, include_lowest=True)

df.pivot_table(index='duration_category', columns='channel', values='duration_minute', aggfunc=['mean', 'median'], observed=False)

mean                                      median  \
channel             FaceBoom MediaTornado     TipTop    organic FaceBoom   
duration_category                                                          
низкая             10.040541     6.923077   8.522388   9.789474     10.0   
средняя            35.178571    36.615385  35.220339  36.828054     31.5   
высокая            81.736842    65.000000  81.562500  86.144578     74.0   

                                               
channel           MediaTornado TipTop organic  
duration_category                              
низкая                     5.5    8.0    10.0  
средняя                   35.5   33.0    35.0  
высокая                   65.0   73.5    83.0

Выводы:
1) Средняя длительность сессии.

В категории "низкая" лидируют два канала - FaceBoom и organic.
В категории "средняя" значения всех каналов примерно одинаковы.
В категории "высокая" - каналы FaceBoom, TipTop и organic показывают хорошие высокие результаты.

2) Медиана.

Почти в каждой категории медианные значения чуть меньше средних, что свидетельствует о наличии небольшого количества очень длинных сессий, которые значительно увеличивают средние показатели.

### Задание №10
 
 **`Добавить столбец - “Длительность сессии” (расчетный). Создать столбец “Категория длительности” (с помощью категоризации). Выделить минимум 3 категории (низкая, высокая, средняя), фильтрацию для уровня длительности выбрать самостоятельно, аргументировать выбор. Создать группировку: средняя, минимальная, максимальная, медианная длительность сессии по категории длительности и устройству `**

Алгоритм вычисления длительности сессии и создания столбца "категория длительности" аналогичен алгоритму в задании №6. Выполняется группировка относительно условий, описанных в задании.

In [67]:
df['duration_minute'] = (df['session_end'] - df['session_start']).dt.total_seconds() / 60

bins = [0, 20, 60, float('inf')]
labels = ['низкая', 'средняя', 'высокая']
df['duration_category'] = pd.cut(df['duration_minute'], bins=bins, labels=labels, include_lowest=True)

df.groupby(['duration_category', 'device'], observed=False)['duration_minute'].agg(avg='mean', min_duration='min', max_duration='max', median='median').reset_index()

duration_category   device        avg  min_duration  max_duration  median
0             низкая  android   9.607143           0.0          20.0    10.0
1             низкая   iphone   8.807692           0.0          20.0     8.0
2             низкая      mac  10.519685           0.0          20.0    10.0
3             низкая       pc   9.519231           0.0          20.0     9.5
4            средняя  android  36.081081          21.0          58.0    35.0
5            средняя   iphone  36.097561          21.0          60.0    34.0
6            средняя      mac  36.454545          21.0          60.0    34.0
7            средняя       pc  37.250000          21.0          60.0    35.0
8            высокая  android  82.851852          61.0         114.0    83.0
9            высокая   iphone  85.520000          61.0         262.0    77.5
10           высокая      mac  83.777778          61.0         154.0    83.0
11           высокая       pc  86.533333          61.0         137.0    79.0

Выводы:
1) Категории "низкая" и "средняя" длительность. 

На всех устройствах наблюдаются схожие показатели.

2) Категория "высокая" длительность. 

Присутсвуют заметные различия в максимальной и медианной длительностях у разных устройств. Самая высокая максимальная длительность наблюдается на iPhone (262 мин), что значительно больше, чем у остальных устройств, при этом медианная длительность - несколько ниже. Это говорит о наличии небольшого количества затяжных сессий.

### Задание №13
 
 **`Добавить столбец - “Длительность сессии” (расчетный). Отфильтровать набор данных. Выбрать только те записи, средняя длительность на которых выше определенного числа (число выбрать самостоятельно). Для отфильтрованного набора данных выполнить группировку: канал привлечения и средняя+медианная длительность`**

Происходит расчет длительности сессии. Выбирается число num=35, относительно которого будут фильтроваться данные. Отфильтрованные данные группируются по условиям, описанным в задании.

In [64]:
df['duration_minute'] = (df['session_end'] - df['session_start']).dt.total_seconds() / 60

num = 35
filtered_df = df[df['duration_minute'] > num]

filtered_df.groupby('channel')['duration_minute'].agg(mean='mean', median='median').reset_index()

channel       mean  median
0      FaceBoom  62.928571    58.5
1  MediaTornado  47.428571    46.0
2        TipTop  58.090909    50.0
3       organic  64.235294    57.0

Выводы:
1) Средняя длительность. 

Самое большое время - у канала organic, самое маленькое - у MediaTornado, что в очередной раз доказывает эффективность первого.

2) Медиана.

Почти в каждой категории медианные значения чуть меньше средних, что свидетельствует о наличии небольшого количества очень длинных сессий, которые значительно увеличивают средние показатели.

### Задание №16
 
 **`Добавить столбец - “Длительность сессии” (расчетный). Отфильтровать набор данных. Выбрать только те записи, средняя длительность на которых ниже определенного числа (число выбрать самостоятельно) + топ 1 устройство по количеству записей + топ 2 канала привлечения по количеству записей . Для отфильтрованного набора данных выполнить группировку: устройство и средняя+медианная+максимальная+ минимальная длительность.`**

Происходит расчет длительности сессии. Выбирается число num=35, относительно которого будут фильтроваться данные. Вычисляется топ 1 устройство и топ 2 рекламных канала. Данные фильтруются. Отфильтрованные данные группируются по условиям, описанным в задании.

In [93]:
df['duration_minute'] = (df['session_end'] - df['session_start']).dt.total_seconds() / 60

num = 35
filtered_df = df[df['duration_minute'] < num]

top_device = filtered_df['device'].value_counts().idxmax()
top_channels = filtered_df['channel'].value_counts()
top_2_channels = top_channels.index[:2]

final_filtered_df = filtered_df[
    (filtered_df['device'] == top_device) &
    (filtered_df['channel'].isin(top_2_channels))
]

final_filtered_df.groupby('device')['duration_minute'].agg(avg='mean', median='median', min_duration='min', max_duration='max').reset_index()

device        avg  median  min_duration  max_duration
0  iphone  14.468182    13.0           0.0          34.0

Вывод:

В выбранном диапазоне (значения меньше 35) длительность варьируется от 0 до 34. Средняя и медианная длительности меньше 35/2, что говорит о наличии большего количества коротких сессий, нежели длинных.

### Задание №21
 
 **`Добавить столбец - “Длительность сессии” (расчетный). Отфильтровать набор данных. Выбрать только те записи, средняя длительность на которых лежит в определенном диапазоне (нижнюю и верхнюю границы выбрать самостоятельно) + топ 1 устройство по максимальному количеству записей+ 2 канала с минимальным количеством записей. Для отфильтрованного набора данных создать сводную таблицу: канал, устройство и минимальная длительность сессии.`**

Произведен расчет длительности сессии. Выбран диапазон фильтрации, данные отфильтрованы. Найдены топовое устровство и 2 самых непопулярных рекламных канала. Создана сводная таблица по условиям задания.

In [95]:
df['duration_minute'] = (df['session_end'] - df['session_start']).dt.total_seconds() / 60

num1 = 28
num2 = 76
filtered_df = df[(df['duration_minute'] > num1) & (df['duration_minute'] < num2)]

top_device = filtered_df['device'].value_counts().idxmax()
channel_counts = filtered_df['channel'].value_counts()
least_2_channels = channel_counts.sort_values(ascending=True).index[:2]

final_filtered_df = filtered_df[
    (filtered_df['device'] == top_device) &
    (filtered_df['channel'].isin(least_2_channels))
]

final_filtered_df.pivot_table(index='channel', columns='device', values='duration_minute', aggfunc='min').reset_index()

device       channel  iphone
0           FaceBoom    29.0
1       MediaTornado    30.0

Вывод:

В двух каналах с минимальным количеством записей длительность сессии на топовом устройстве (iPhone) практически одинаковое.

### Задание №25
 
 **`Добавить столбец - “Длительность сессии” (расчетный). Создать столбец “Категория длительности” (с помощью категоризации). Выделить минимум 3 категории (низкая, высокая, средняя), фильтрацию для уровня длительности выбрать самостоятельно, аргументировать выбор. Отфильтровать набор данных. Выбрать только высокую и низкую длительность сессии + 2 самых не популярных канала (по количеству записей). Создать сводную таблицу: средняя, медианная длительность сессии по каналу и устройству.`**

Рассчитывается длительность сессии, создается столбец "категория длительности" аналогично предыдущим заданиям. Из набора данных вычленяются те записи, которые соответсвуют низкой и высокой категории длительности. Аналогично прошлому заданию находятся 2 самых непопулярных рекламных канала. Данные фильтруются, создается сводная таблица по условиям задания.

In [96]:
df['duration_minute'] = (df['session_end'] - df['session_start']).dt.total_seconds() / 60

bins = [0, 20, 60, float('inf')]
labels = ['низкая', 'средняя', 'высокая']
df['duration_category'] = pd.cut(df['duration_minute'], bins=bins, labels=labels, include_lowest=True)

filtered_df = df[df['duration_category'].isin(['низкая', 'высокая'])]
channel_counts = filtered_df['channel'].value_counts()
least_2_channels = channel_counts.sort_values(ascending=True).index[:2]

final_filtered_df = filtered_df[filtered_df['channel'].isin(least_2_channels)]

final_filtered_df.pivot_table(
    index=['channel', 'device'],
    values='duration_minute',
    aggfunc=['mean', 'median']
).reset_index()

channel   device            mean          median
                         duration_minute duration_minute
0  MediaTornado  android       15.750000            10.0
1  MediaTornado   iphone        4.923077             3.0
2  MediaTornado      mac        9.250000             9.5
3  MediaTornado       pc        9.000000             9.0
4        TipTop  android       24.357143            14.0
5        TipTop   iphone       19.622222             6.0
6        TipTop      mac       25.444444            12.0
7        TipTop       pc       32.333333            10.5

Выводы:
1) MediaTornado.

Наиболее долгие сессии у пользователей на Android (средне и медиана — значительно выше, чем у других устройств). Пользователи iPhone показывают наименьшую продолжительность как по среднему, так и по медиане, что может указывать на более короткое время сессии на iPhone. Продолжительность на Mac и PC примерно совпадает, немного выше у Mac.

2) TipTop.

Наиболее продолжительные сессии у пользователей на PC. Устройства Mac и Android показывают схожие средние значения, но медиана у Mac чуть выше, что говорит о большей стабильности или большем числе более длинных сессий. Пользователи iPhone демонстрируют меньшую продолжительность, особенно по медиане (6 мин), что отличается от средней — вероятно, присутствует несколько очень длинных сессий, повышающих среднее значение.